[data][llm] Expose logprobs support in Ray Data LLM

nrghosh opened the pull request:
## Description
Exposes logprobs support in Ray Data LLM by extracting logprobs and prompt_logprobs from vLLM outputs and including them in the output rows.

## Changes

- Add `logprobs` and `prompt_logprobs` fields to `vLLMOutputData` model
- Extract logprobs from `output.outputs[0].logprobs` in `from_vllm_engine_output`
- Extract prompt_logprobs from `output.prompt_logprobs` in `from_vllm_engine_output`
- Convert vLLM's `Logprob` dataclass instances to serializable dicts using `dataclasses.asdict()`

## Testing

Added unit tests verifying:
- Extraction of logprobs for generated tokens (including multiple logprobs per token)
- Extraction of prompt_logprobs (including None entries)
- Proper handling when logprobs are not requested

## Related issues
Closes https://github.com/ray-project/ray/issues/58894 

## Additional information

nrghosh committed:
wip - expose logprobs support in Ray Data LLM

Extract and surface logprobs from vLLM outputs. Previously, logprobs
could be requested in sampling_params but were not returned in output
rows. This adds logprobs and prompt_logprobs fields to vLLMOutputData
and extracts them from vLLM's RequestOutput.

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - add tests for logprobs extraction

Add unit tests for logprobs and prompt_logprobs extraction from vLLM
outputs, including cases with multiple logprobs per token and None
values.

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

kouroshHakha commented on python/ray/llm/_internal/batch/stages/vllm_engine_stage.py:
```diff
@@ -111,6 +119,28 @@ def from_vllm_engine_output(cls, output: Any) -> "vLLMOutputData":
+            if output.outputs[0].logprobs is not None:
+                data.logprobs = [
+                    {
+                        token_id: dataclasses.asdict(logprob)
```
What are the possible types for logprob object? Can it be pydantic as well as dataclass (think about the future changes that could happen and the diff between sglang and vllm). I am afraid using dataclasses.asdict() might overfit to todays version

nrghosh replied:
yeah good q

* vLLM uses dataclass (and they note that it might change in the future (https://github.com/vllm-project/vllm/blob/main/vllm/logprobs.py#L9-L11) 
* SGLang uses BaseModel 

so I've added cases / flexibility to `_convert_logprob_dict` to handle those (+ future formats)

kouroshHakha commented on python/ray/llm/_internal/batch/stages/vllm_engine_stage.py:
```diff
@@ -111,6 +119,28 @@ def from_vllm_engine_output(cls, output: Any) -> "vLLMOutputData":
+                    if logprob_dict is not None
+                    else None
+                    for logprob_dict in output.prompt_logprobs
+                ]
```
make this part a utility and reuse between prompt_logprobs and logprobs

nrghosh replied:
refactored to `_convert_logprob_dict`

kouroshHakha commented on python/ray/llm/tests/batch/gpu/stages/test_vllm_engine_stage.py:
```diff
@@ -439,5 +440,97 @@ class AnswerModel(BaseModel):
+
+    logprobs = [
+        {
+            123: Logprob(logprob=-0.5, rank=1, decoded_token="hello"),
```
question, what is rank here? rank of the TP workers? or some other rank?

nrghosh replied:
bit of a naming collision - `rank` in this case is the vocab rank of the token, not related to TP workers or anything like that. 

> From vllm docs: "The vocab rank of chosen token (>=1)"
ex rank=1 means it's the most probable token (top-1)

This comes from vLLM's sampler by counting how many tokens have higher logprobs than the chosen token. So it's just a representation of model confidence.

nrghosh committed:
wip - cleanup tests

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Refactor logprobs conversion to use duck-typed utility function

- Extract _convert_logprob_dict() helper to eliminate code duplication
- Support dataclass, Pydantic, and dict formats for future compatibility
- Raise explicit TypeError for unsupported types

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh reviewed (commented):
premerge gpu tests failing

```
=========================== short test summary info ============================
--
[2025-11-25T02:53:59Z] FAILED python/ray/llm/tests/batch/gpu/stages/test_vllm_engine_stage.py::test_vllm_wrapper_generate
[2025-11-25T02:53:59Z] FAILED python/ray/llm/tests/batch/gpu/stages/test_vllm_engine_stage.py::test_vllm_wrapper_lora
[2025-11-25T02:53:59Z] FAILED python/ray/llm/tests/batch/gpu/stages/test_vllm_engine_stage.py::test_vllm_wrapper_json
[2025-11-25T02:53:59Z] ==================== 3 failed, 6 passed in 78.37s (0:01:18) ====================

```

nb: should be fixed by https://github.com/ray-project/ray/pull/58980

nrghosh committed:
Merge branch 'master' into nrghosh/expose-logprobs

ruisearch42 reviewed (approved):
LGTM
